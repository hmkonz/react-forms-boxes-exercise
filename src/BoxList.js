import React, {useState} from "react";
import Box from "./Box";
import NewBoxForm from "./NewBoxForm";


const BoxList = () => {
   const [boxes, setBoxes] = useState([]);

   const addBox = ({height, width, backgroundColor}) => {
        // returns a new array with all the existing boxes as well as a new object with id, height, width and background color of new box
        setBoxes(boxes =>[...boxes, {height, width, backgroundColor}])
   };

   const remove = id => { 
    // 'id' passed in to remove() is the id of the box to be removed

    // filter() interates over the in-state 'boxes' array and passes in each id through the callback function. When box.id !== id returns false (means box.id===id), that 'box' element from the array is not added to the new array created. SetBoxes is now passed that new array and changes the stored state of 'boxes' to this new array.
        setBoxes(boxes => boxes.filter(box => box.id !== id));
    }

    return (
        <div>
            <NewBoxForm addBox={addBox} />
            {boxes.map(box => (
                <Box 
                    key={box.id}
                    id={box.id} 
                    height={box.height} 
                    width={box.width} 
                    backgroundColor={box.backgroundColor} 
                    handleRemove={remove} 
                />
            ))}
             
        </div>    
    
  )
}

export default BoxList;







