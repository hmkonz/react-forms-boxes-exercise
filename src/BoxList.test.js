import '@testing-library/jest-dom/extend-expect';
import React from "react";
import { render, fireEvent } from "@testing-library/react";
import BoxList from "./BoxList";

// smoke test
it("renders without crashing", function() {
    render (<BoxList />);
});

// Snapshot test
it("matches snapshot", function() {
    const {asFragment} = render(<BoxList />);
    expect (asFragment()).toMatchSnapshot();
});

it("can add a new box", function() {
    const {getByLabelText, queryByText, getByPlaceholderText} = render (<BoxList />);

    // get the input values of Height, Width and Background Color as well as the button from the form
    const heightInput = getByLabelText("Height");
    const widthInput = getByLabelText("Width");
    const backgroundInput = getByLabelText("Background Color");
    const button = queryByText("Add a new box!");

    // no boxes added yet so confirm there isn't a button with an "X" which is added with every box
    expect(queryByText("X")).not.toBeInTheDocument();

    // Add a box (by mock filling out the form)
    // Mocks what will be filled out in the backgroundInput, widthInput and heightInput inputs of the form and updates the state of "boxes"
    fireEvent.change(backgroundInput, { target: { value: "Blue" } });
    fireEvent.change(widthInput, { target: { value: "5" } });
    fireEvent.change(heightInput, { target: { value: "5" } });
    //mocks submitting the button 
    fireEvent.click(button);

    // when box is added, expect to see button with an 'X" on it
    expect(queryByText("X")).toBeInTheDocument();
  
    // after box is added, inputs are cleared and expect to see placeholders in inputs
    const heightPlaceholder = getByPlaceholderText("Height");
    expect(heightPlaceholder).toBeInTheDocument();
});

it("can remove a box", function() {
    const {getByLabelText, queryByText} = render (<BoxList />);

    // get the inputs of Height, Width and Background Color as well as the button from the form
    const heightInput = getByLabelText("Height");
    const widthInput = getByLabelText("Width");
    const backgroundInput = getByLabelText("Background Color");
    const button = queryByText("Add a new box!");

    // add a box (by mock filling out the form)
    fireEvent.change(backgroundInput, { target: { value: "Blue" } });
    fireEvent.change(widthInput, { target: { value: "5" } });
    fireEvent.change(heightInput, { target: { value: "5" } });
    //mocks submitting the button 
    fireEvent.click(button);

    // click the remove button ('X') and the box should be gone
    fireEvent.click(queryByText("X"));
    expect(queryByText("X")).not.toBeInTheDocument();

});


