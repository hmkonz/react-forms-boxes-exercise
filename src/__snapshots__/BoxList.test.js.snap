// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <div>
      <form>
        <div>
          <label
            for="height"
          >
            Height
          </label>
          <input
            id="height"
            name="height"
            placeholder="Height"
            type="text"
            value=""
          />
        </div>
        <div>
          <label
            for="width"
          >
            Width
          </label>
          <input
            id="width"
            name="width"
            placeholder="Width"
            type="text"
            value=""
          />
        </div>
        <div>
          <label
            for="backgroundColor"
          >
            Background Color
          </label>
          <input
            id="backgroundColor"
            name="backgroundColor"
            placeholder="Background Color"
            type="text"
            value=""
          />
        </div>
        <button>
          Add a new box!
        </button>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
